### ISSUE TITLE
Improve usability of `make cluster` for typical dev workflow

### ISSUE THREAD

**Issue by cjcjameson** [greenplum-db/gpdb#2422]
We would like to improve the "demo cluster" as invoked by `make cluster` from gpAux/gpdemo.

_in rough order, from low-hanging fruit to gnarly._

1. [x] Expose `make cluster` as a top-level make target. Typical dev workflows currently `cd` or use `make -C` to call it.

1. [ ] Check earlier / fail faster if the localhost networking configuration (hosts file, passwordless SSH, etc.) will keep the demo cluster from coming up. Try to have better error messages. Example of pain: https://github.com/greenplum-db/gpdb/issues/1622. Example of a place where a workaround was needed: https://github.com/greenplum-db/gpdb/pull/2452

1. [ ] Address the underlying issues that require configuration of networking on localhost.

1. [ ] `make cluster` can fail if you don't have certain environment variables available over SSH. There are various ways to make it work, but we can fix this across setups, perhaps by installing with RPATH when we create postgres

1. [ ] `make distclean` will likely disrupt further management operations upon the demo cluster. But `make distclean` shouldn't be used nearly as much as some engineers currently might use it. Address this by enabling devs to stamp out any superstitions about `make distclean` and just rely on `make clean`.

1. [ ] one way that `make distclean` could disrupt the cluster is if it removes `Makefile.global`. We should change the make files so that the demo cluster makefile doesn't depend on `Makefile.global`

1. [ ] Depending on the install prefix, there are sometimes issues with libraries' location. @hardikar for more detail.

1. [ ] `make cluster` is slow relative to manually orchestrating a local cluster spin-up with lower-level cluster instantiation primitives (e.g. `pg_ctl`)

Based on these improvements, we'd like to improve the top-level readme's Building GPDB section such that it can read:

```bash
./configure
make
make install
make cluster
# calls to various tests
```

We assume that the configure flag complexity is reduced in #2420 -- out of scope for this issue

**Comment by Eulerizeit** [comment]
@cjcjameson currently @danielgustafsson is woking with the gpdemo DB for pg_upgrade testing and then the C&S team ( @jimmyyih  ) are going to pick up that work and add the binary compatibility tests. This seems like a good idea that can possibly be rolled into some of that work.

**Comment by danielgustafsson** [comment]
Re gpdemo being slow compared to pg_ctl: since the output of `make cluster` is what is used for ICW, I think it makes sense that we use what we expect a normal user to use. gpdemo is indeed a slow beast, but we should fix that rather than change the underlying primitives.

For testing pg_upgrade, it's important that we can have two separate clusters running completely in parallel, see https://github.com/greenplum-db/gpdb/pull/2386/commits/897e93e6cdf9b9b1a093da7c7ef6685979d48bd7.

**Comment by skahler-pivotal** [comment]
I don't like the demo environment as a top level "make cluster" command. Maybe "make democluster" would be more appropriate. My concern would be that the command at that level would imply at that level you are making "the cluster" or your dedicated long running instance of gpdb.

**Comment by danielgustafsson** [comment]
> On 16 May 2017, at 18:05, Scott Kahler <notifications@github.com> wrote:
> 
> I don't like the demo environment as a top level "make cluster" command. Maybe "make democluster" would be more appropriate. My concern would be that the command at that level would imply at that level you are making "the cluster" or your dedicated long running instance of gpdb.

I completely agree with this, democluster is a better name.

**Comment by cjcjameson** [comment]
Sweet! 1 of 8 down. We'll try to keep getting to this next week, bit by bit.

**Comment by cjcjameson** [comment]
The prioritization of open action items above still seems basically correct. In connection with the first two open action items (both about networking to localhost), I think that recent struggles with IPv6 and macOS Sierra may be connected. As laid out above, there are improvements to be made in failing faster if the cluster won't work, and also improvements to be made to GPDB itself.

**Comment by cjcjameson** [comment]
@professor if you're able to reproduce the IPv6 issue on Sierra, feel free to open a new issue and link back to this overarching issue. I think various different people have good ideas about different ways to solve it, good to have a separate conversation. #1622 and #2452 have some prior history, but others' experiences are very recent.

**Comment by Eulerizeit** [comment]
closing given a lack of consensus
